src/rules/no-shallow-imports.js

```diff
@@ -0,0 +1,104 @@
+import fs from 'fs';
+import minimatch from 'minimatch';
+import docsUrl from '../docsUrl';
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Prevent the use of shallow imports',
+      recommended: true,
+      url: docsUrl('no-shallow-imports'),
+    },
+    schema: [
+      {
+        oneOf: [
+          {
+            type: 'object',
+            properties: {
+              allow: {
+                type: 'array',
+                items: {
+                  type: 'string',
+                },
+              },
+            },
+            additionalProperties: false,
+          },
+        ],
+      },
+    ],
+  },
+  create: context => {
+    return {
+      ImportDeclaration: node => maybeReportShallowImport(node, context),
+    };
+  },
+};
+
+const maybeReportShallowImport = (node, context) => {
+  const filenamePath = context.getFilename();
+
+  const options = context.options[0] || {};
+  const allowRegexps = (options.allow || []).map(p => minimatch.makeRe(p));
+
+  const doesFileMatchAllowRegexps = allowRegexps.some(regex => regex.test(filenamePath));
+
+  // If the file does not match the regexps and is a shallow import, then report it
+  if (!doesFileMatchAllowRegexps && isShallowImport(node, context)) {
+    return context.report({
+      node,
+      message: 'Please deep import!',
+    });
+  }
+};
+
+const isCalledIndex = filename => filename.slice(0, 5) === 'index';
```

Comment:
```suggestion
const isCalledIndex = filename => filename.split('.')[0] === 'index';
```

Think this might be slightly better because otherwise:

```js
isCalledIndex('index-lorem-ipsum.tsx') 
```
will be `true` where as it should be `false`.

Reply:
```suggestion
const isCalledIndex = filename => path.basename(filename, path.extname(filename)) === 'index';
```

Reply:
@ljharb `path.basename(filename, path.extname(filename)) === 'index'` would evaluate to `false` if the filename is `index.test.js` for example

`filename.split('.')[0] === 'index'` might be safer - what do you think?

Reply:
Sure, that sounds good